VisualPinball.Unity/VisualPinball.Unity.Editor/Layers/LayerHandler.cs

```diff
@@ -1,35 +1,300 @@
+			}
+			if (!_layers.ContainsKey(item.EditorLayerName)) {
+				_layers.Add(item.EditorLayerName, new List<MonoBehaviour>());
+			}
+			_layers[item.EditorLayerName].Add((MonoBehaviour)item);
+		}
+
+		private void RebuildTree()
+		{
+			TreeRoot.Children.Clear();
+
+			// init with root element
+			if (_tableBehavior != null && _tableBehavior.Table != null) {
+
+				// table node
+				var tableItem = new LayerTreeElement(_tableBehavior.Table) { Id = 0 };
+				TreeRoot.AddChild(tableItem);
+
+				var layerCount = 1;
+				foreach (var pair in _layers.OrderBy(key=> key.Key)) {
+
+					// layer node
+					var layerItem = new LayerTreeElement(pair.Key) { Id = layerCount++ };
+					tableItem.AddChild(layerItem);
+
+					foreach (var item in pair.Value.OrderBy(behaviour => behaviour.name)) {
+						if (item is ILayerableItemBehavior layeredItem) {
+							layerItem.AddChild(new LayerTreeElement(layeredItem) { Id = item.gameObject.GetInstanceID() });
+						}
+					}
+				}
+			}
+
+			TreeRebuilt?.Invoke();
+		}
+		#endregion
+
+		#region Layer renaming
+		/// <summary>
+		/// Update the layer name of an ILayerableItemBehavior, managing Undo
+		/// </summary>
+		/// <param name="item">the ILayerableItemBehavior to modify</param>
+		/// <param name="layerName">the new layer name</param>
+		private void SetLayerNameToItem(ILayerableItemBehavior item, string layerName)
+		{
+			if (item.EditorLayerName != layerName) {
+				if (item is MonoBehaviour behaviour) {
+					Undo.RecordObject(behaviour, $"Item {behaviour.name} : Change layer name from {item.EditorLayerName} to {layerName}");
+				}
+				item.EditorLayerName = layerName;
+			}
+		}
+
+		/// <summary>
+		/// Will update recursively the BiffData EditorLayerName of all provided elements with a new layer name
+		/// </summary>
+		/// <param name="elements">an array of LayerTreeElement to parse</param>
+		/// <param name="layerName">the new layer name</param>
+		private void SetLayerNameToItems(LayerTreeElement[] elements, string layerName)
 		{
-			return !SceneVisibilityManager.instance.IsHidden(gameObj);
+			foreach (var element in elements) {
+				if (element.Item != null) {
+					SetLayerNameToItem(element.Item, layerName);
+					if (element.HasChildren) {
+						SetLayerNameToItems(element.GetChildren<LayerTreeElement>(), layerName);
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Callback when LayerTreeView as validated a layer rename
+		/// </summary>
+		/// <param name="element">the the renamed layer TreeElement</param>
+		/// <param name="newName">the new validated name</param>
+		internal void OnLayerRenamed(LayerTreeElement element, string newName)
+		{
+			if (element.LayerName == newName) {
+				return;
+			}
+
+			//Check if there is not already a layers with thr same name
+			if (_layers.ContainsKey(newName)) {
+				EditorUtility.DisplayDialog("Visual Pinball", $"There is already a layer named {newName}.\nFind another layer name.", "Close");

```

Comment:
VPX allows duplicate layer names. I agree it doesn't make much sense in terms of usability, but that means we can get duplicate layer names from imports, which, as far as I can see, will break your `_layers` logic. Maybe use the layer ID to index?

Reply:
Right, I'll check that

Reply:
Hmm what do you mean by reloading? Save, close and open? How are the layers written?

In general the identifier seems to be the ID, not the name. Why not use the ID then? Does the trigger other problems?

Reply:
VPX7 is using LANR data first when loading items. If it's empty it creates a "Layer_num"  layer name using LAYR data and set LANR data with it. On next save, LAYR shouldn't be used anymore.
You can see it at the end of PinTable::LoadGameFromStorage in VPX7 project.
I did the same when initializing layer names when opening the editor.
Tthe LANR item's data is the only place where layer names are stored, so having 2 layers with the same name last time you saved your table will lead to layer merging the next time you'll load it.
VPX6 tables are not using LANR at all, so i'll create "Layer_num" layers in this case (as VPX7 does).

Reply:
If you're saying VP doesn't save multiple layers with the same name, then it's okay.